server/src/internalClusterTest/java/org/elasticsearch/repositories/IndexSnapshotsServiceIT.java

```diff
@@ -0,0 +1,358 @@
+        createIndex(indexName, indexName2);
+        SnapshotInfo lastSnapshot = null;
+        int numSnapshots = randomIntBetween(5, 25);
+        for (int i = 0; i < numSnapshots; i++) {
+            if (randomBoolean()) {
+                indexRandomDocs(indexName, 5);
+                indexRandomDocs(indexName2, 10);
+            }
+            final List<String> snapshotIndices = randomSubsetOf(indices);
+            final SnapshotInfo snapshotInfo = createSnapshot(repoName, String.format(Locale.ROOT, "snap-%03d", i), snapshotIndices);
+            if (snapshotInfo.indices().contains(indexName)) {
+                lastSnapshot = snapshotInfo;
+            }
+        }
+
+        if (useBwCFormat) {
+            // Reload the RepositoryData so we don't use cached data that wasn't serialized
+            assertAcked(clusterAdmin().prepareDeleteRepository(repoName).get());
+            createRepository(repoName, "fs", repoPath);
+        }
+
+        final Optional<ShardSnapshotInfo> indexShardSnapshotInfoOpt = getLatestSnapshotForShard(repoName, indexName, 0);
+        if (lastSnapshot == null) {
+            assertThat(indexShardSnapshotInfoOpt.isPresent(), equalTo(false));
+        } else {
+            assertThat(indexShardSnapshotInfoOpt.isPresent(), equalTo(true));
+
+            final ShardSnapshotInfo shardSnapshotInfo = indexShardSnapshotInfoOpt.get();
+
+            final ClusterStateResponse clusterStateResponse = admin().cluster().prepareState().execute().actionGet();
+            final IndexMetadata indexMetadata = clusterStateResponse.getState().metadata().index(indexName);
+            final String indexMetadataId = IndexMetaDataGenerations.buildUniqueIdentifier(indexMetadata);
+            assertThat(shardSnapshotInfo.getIndexMetadataIdentifier(), equalTo(indexMetadataId));
+
+            final SnapshotInfo snapshotInfo = shardSnapshotInfo.getSnapshotInfo();
+            assertThat(snapshotInfo, equalTo(lastSnapshot));
+            assertThat(snapshotInfo.state(), equalTo(SnapshotState.SUCCESS));
+            assertThat(snapshotInfo.indices().contains(indexName), equalTo(true));
+        }
+    }
+
+    public void testGetShardSnapshotWhileThereIsARunningSnapshot() throws Exception {
+        final String fsRepoName = randomAlphaOfLength(10);
+        createRepository(fsRepoName, "mock");
+
+        createSnapshot(fsRepoName, "empty-snap", Collections.emptyList());
+
+        final String indexName = "test-idx";
+        createIndexWithContent(indexName);
+
+        blockAllDataNodes(fsRepoName);
+
+        final String snapshotName = "snap-1";
+        client().admin().cluster().prepareCreateSnapshot(fsRepoName, snapshotName).setIndices(indexName).setWaitForCompletion(false).get();
+
+        waitForBlockOnAnyDataNode(fsRepoName);
+
+        assertThat(getLatestSnapshotForShard(fsRepoName, indexName, 0).isEmpty(), equalTo(true));
+
+        unblockAllDataNodes(fsRepoName);
+
+        assertBusy(() -> {
```

Comment:
You could just use `org.elasticsearch.snapshots.AbstractSnapshotIntegTestCase#awaitNoMoreRunningOperations()` here I think to wait for the snapshot to finish and avoid the busy assert?
Or even cleaner, just use `waitForCompletion(true)` and just keep track of the `ActionFuture<CreateSnapshotResponse>` from the snapshot request (use `execute` instead of `get` for that) and then just call `org.elasticsearch.snapshots.AbstractSnapshotIntegTestCase#assertSuccessful(org.elasticsearch.action.ActionFuture<org.elasticsearch.action.admin.cluster.snapshots.create.CreateSnapshotResponse>)` on it once you unblocked the data nodes. Gives you much more deterministic behavior and should run faster as well :)